Initial SRE Maturity Model content (#4)
operate-first/SRE: sre_maturity.md

# SRE Maturity Model

## Phase 1: The ‘Crawl’ Phase

### Crawl Phase Overview

You’re starting from zero, with a net new team.

> You may only have a small SRE team, or an SRE function within your development team.

Reviewer: I'm not sure that the size of the SRE team necessarily correlates with some of the other measures. In the RHEL org, even small (non-managed-service) teams which are responsible for running their own service (think CI services for internal customers etc) are expected to provide a reliable service, managed according to the later stages of the model.

Author: Agreed.
Also, there are other personas, like existing ops or devops teams, that could be at the early phases of maturity in some areas.
I'm going to rearrange the PR so there are phases of maturity in different areas rather than talking about overall maturity

Your team needs room to practice processes, improve them and automate things.
Without this room, your team will mature more slowly and get bogged down in just keeping the service running day to day.
This can lead to scaling problems, both for the service and the team.
It isn't necessary to formally measure how much time is spent on keeping things running vs. feature development or automation.
That can come later, however you are familiar with the concept of Toil.

## Phase 2: The ‘Walk’ Phase (WIP)

### Walk Phase Overview

Your team is well established, has working processes and is familiar with SRE principles.
A lot of those principles are baked into their processes & daily work.
In this phase your team is building on the foundations of observability and starting to reflect on their learnings about the service.
Feedback loops, automation and 'shifting left' with different types of testing are the main theme of this phase.
Your team is capable of scaling the service and number of users in a profitable way.

### Walk Phase Characteristics

* SLOs - SLOs have been iterated on, and you have more than 1 meaningful SLO. You understand and track error budgets

> * Observability

Reviewer 2: I think in this phase with a well established team that understands common SRE practices SLOs would start to appear for their service(s) as a way of placing measuring sticks around the effectiveness of scaling efforts. Error budgets wouldn't be tracked, at least not formally.

The SLO might only be one around service availability and it may not even be formally tracked at the start of this phase (as one is first learning to walk). Towards the end of the phase, as the team transitions to "running," the they may start to wonder how to enforce that availability SLO and other new ones they want to put into place. I can imagine a conversation between an engineer and product owner where a PO is asking why customers are complaining about the service and the engineer saying that they never have time to work on improving availability, and it dawning on the PO that something like an error budget is needed.

* Chaos Testing (possible candidate for Phase 1, but what would that look like? Ad-hoc testing?)
* Starting to get Exec support (improvment on 'buy-in')
* Start measuring Toil more formally & management track it
  * Develop relationship with the upstream (eng and sre, or sre & upstream product) for managing toil
* Improving the architecture for lower cost to serve (Is there a prereq here - calculating & knowing the cost to serve)
* DR & Incident Rehearsals (Identifying failure modes could be a prerequisite in phase 1)

## Phase 3: The 'Run' Phase (WIP)

### Run Phase Overview

Your team has found a balance between keeping things running and building reliability into the service.
They have the ability to scale the service without having to increase team size.
There is sufficient confidence in the service that your team looks at 'shifting right' with things like canary releases and production testing.
Automation and observability are mature enough to start looking at concepts like AI Ops.

### Run Phase Characteristics

* Observability correlation beteween logs, metrics, alerts, events etc..
* AI Ops
  * Needs everything underneath it first (all the signals)
* Chaos Engineering

> * SLOs - Well established SLO Review process

Reviewer 2: Continuing with the thread of SLO review in each phase, I think this is where a late phase 3 team will be.

# SRE Maturity Model

## Phase 1: The ‘Crawl’ Phase

### Overview

You’re starting from zero, with a net new team.
You may only have a small SRE team, or an SRE function within your development team.
You and your team are technically capable of surviving.
You are probably providing a ‘white glove’ experience to your users at this stage.

### Characteristics

#### People and Culture

You have a strong SRE voice in your team, someone who advocates for reliability features.
They establish people focused mantras like "It's OK to fail", "Hope is not a strategy" and "Trust by default & extend on it".
Communication is important in your team and you hire diverse people based on soft skills & leadership rather than specific technologies.

> 

Reviewer 3: You may also have folks from previous "disciplines". For example, sys admins, or experienced developers. Or, even some folks who have experience in "devops". They can bring great skills, experience and value, but they may need to be re-trained (or even un-trained) into the SRE paradigm.

Author: ```suggestion
You may also have folks from previous "disciplines". For example, sys admins, or experienced developers. Or, even some folks who have experience in "devops". They can bring great skills, experience and value.
You are likely to see resistance to change when making this transition with existing team members, so retraining (or even unlearning) can help. However, there may be inevitable difficult personnel decisions to be made.
```

Thanks @brizrobbo 
I've taken most of what you've said and changed the end slightly to incorporate some feedback from Narayanan (paraphrased here).
How does this sound?

At this phase, having buy-in from executive level leadership is essential.
Without this it will be very difficult to get their support later, and eventually their engagement.

The product manager for you service should have a good understanding of SRE.
They should help you advocate for service features that make the service more reliable.

> #### Observability

Author: @jeremyeder @mmazur @lisa 
Probably something for future PRs, but can I get your thoughts on a 'Pitfalls' section in each phase.
It would highlight lessons learned where we've tried something and it didn't work or it slowed things down.

For example, in kafka, a lot of alerts were defined very early in the service journey. People familiar with kafka & strimzi came up with lots of great ways to detect that something may be wrong by querying various metrics. 
This led to a few pain points.
* lots of alerts being lowered to 'warning' severity, firing regularly in production and essentially being ignored. (A good alerting consistency model & policy can help with this)
* The sheer volume of alerts tied to a hard requirement that came later to add SOPs for all alerts (why have an alert that isn't actionable?) led to a big slowdown while playing catchup.

There's another similar pain point where alerts defined in an upstream project/product are taken as is. This can lead to false positives as thresholds may be different or the alert use case is a little different than how you're using the upstream project/product.

Reviewer 2: Maybe each phase has a section on observability and for the "Crawl phase" one characteristic is an excess of alerts that are later silenced, ignored, or have thresholds tuned to page less often. Later phases can refine the practice by pruning "vanity alerts" and honing in on the alerts which directly support specific SLOs.

Author: That makes sense @lisa 
I'm going to rework the model to support this approach.

The foundations for monitoring and observing your service are in place.
Being able to measure something meaningful about your service is important, and supplements any feedback you get from initial users.
A common way to do this is via metrics exposed by your service.

> You have a simple Service Level Objective (SLO) for you service to start out with.

Author: This was a strong suggestion from a PM to have this in phase 1

Author: @lisa I'd appreciate your thoughts on the SLO pieces in the different phases

Reviewer 2: I can have a look this afternoon.

Reviewer 2: My gut says that at the "Crawl Phase" new SRE teams will not have any SLOs. I'd expect more like a traditional measure and alert on everything approach that leads to pager fatigue and a shotgun blast of alerts every time something goes wrong. I recognise that there's a desire to have an SLO in phase 1 and I think it's a great idea if it was, but if we're describing "bands" where existing teams may be, then I don't think it's a reality that they have any SLOs in place.

Alerting is likely to be reactive in nature, put in place because an incident occurred and caused component X to do some behaviour that could be measured. When component X's metric is exhibiting the behaviour and meets some threshold decided after the incident, someone is paged. Maybe that's memory usage, CPU time, a queue's size, disk space, number of replicas, pod count or something else, but I don't think they'd be called SLOs.

I think SLOs would start to be considered later in the maturity model, likely in the "Walk Phase" as the team becomes more comfortable with common SRE practices. I'll add notes in the other sections.

Reviewer 4: Lisa's right in that most people start by over-alerting first (and I agree with that list of metrics that people typically start with) and then dial it down in some areas and up in others as they learn what's really important.

In Red Hat Developer we are now getting Managed Services teams to write SLOs very early on, from the first preview release, which is really cool, but that's the first time I've seen it done that way.

It doesn't have to be completely representative of what users expect from your service, but it forms the building blocks of an SLO review process and improving on them in time.

> Starting with a lower SLO than where you want to get to sets expectations early.

Reviewer 5: Here we might mention a "graduation" process from "informing" to "enforcing" SLO.  This might take a few sprints.

Author: ```suggestion
Starting with a lower SLO than where you want to get to sets expectations early. There's a graduation process with SLOs from "informing" to "enforcing". You can use SLOs to inform decisions in this early stage. As they get refined and wider awareness & support, you can think about enforcement of an error budget policy.
```
@jeremyeder This sound resonable?

> #### “Admin” access to the service

Reviewer 5: This is being called out because there are SRE teams somewhere who don't have this?

Author: @jeremyeder Good question.
This came up as a suggestion on 1 of the maturity calls as something for phase 1 that you absolutely need for your SRE team.

How about this?
Building on the comment from @billmvt below, the title of the section could change to something security focused. e.g. "Restricted access to the service"

Phase 1 proposes "production access restricted to a tightly scoped set of individuals".
Ad hoc security scans or pen tests would happen here too.

Phase 2 is 'shifting left' with scans in source code & 'shifting right' with regular or ongoing production scans. 
Couple this with 'security impact assessments' being considered at the design phase of features and architecture decisions.
Auditing of user access and having security compliance is a must in this phase.

Phase 3, I'm not sure but something  around "Automated Policy Enforcement".
I don't know what that means just yet.

Author: I've pushed some changes to the structure. There's a section for 'Security' phases of maturity now.

Your team has the level of access required to fix production problems.
If not, you can’t expect them to debug and fix problems in a timely manner.
Being overly controlling of access shouldn’t be a priority at this phase.
That can come later with things like audited access, permissions escalation & tiered access as you mature the debugging experience of your service.

> You should trust your team and assume positive intent.

Reviewer 6: Just a consideration: all enterprise customers expect cloud services to have some compliance certification relevant to their industry. And every compliance framework (SOC-2, ISO 27001, PCI-DSS, HIPAA / HITRUST, FedRAMP / NIST 800-53) requires access policies that are defined by management and enforced technically. 

Additionally, most compliance frameworks demand personnel background checks for individuals with privileged access to the environment. The more individuals with privileged access, the more background checks you have to run, which Red Hat does not do enterprise-wide today. 

I might refine this guidance to be something like: "Don't bother getting granular with authz policies to start with, but do start with production access restricted to a tightly scoped set of individuals."

#### On Call, Escalations and Incidents

You have a team with an on call schedule.
It doesn’t have to be 24/7. It doesn’t have to be automated.
But there must be someone responsible to respond when things go bad.
Setting expectations at this early stage with a reduced SLO means you can tailor the schedule to more reasonable times for your team.

> Following on from this, you have an escalation process. A way to be notified of problems. 

Reviewer 5: I'm tripping over the "A way to be notified of problems" sentence, I don't think it's specific enough.  What is it referring to that's different than what was said in earlier bullets of this section?

Author: ```suggestion
Following on from this, you have an escalation process. 
```
Let's remove it.
I think I wrote it before the bullets above and now it doesn't flow.

It should be possible to raise a problem with the SRE team and the people that build the service, like engineers, QE, writers.
This can be a mailing list, slack channel, issue system, whatever.
Formalise it and share it.

> If things go really bad, you write up an RCA.

Reviewer 5: make sure to expand all acronyms the first time they're used.  also this jumps straight to RCA without first saying that theres an incident management process and an incident has been triggered.

Reviewer 3: Also, some teams don't call them RCAs, but "post mortem reviews".

Author: ```suggestion
If things go really bad, you have an Incident Management Process and do a Root Cause Analysis (RCA) (sometimes called a Post Mortem or Post Incident Review).
```
@jeremyeder @brizrobbo Reworded.
I'm hesitant to go into detail on Incident Management Process here as I think there's an opportunity to publish a slimmed down generic version of the process used in SD today.
If you don't think that's unreasonable I can capture it under https://issues.redhat.com/browse/SIGSRE-50 (There's one in there already for an RCA template)

You should be starting to build up a library of RCAs that can feed into doing more proactive things as you mature e.g. Incident rehearsals.
Each RCA includes a 'blameless' post mortem session with the people involved while the incident is still fresh in their head.

> This post mortem feeds important reliability improvements back to the team.

Reviewer 5: Add emphasis here that the value of post mortems is in the changes they recommend. Lack of follow-up here is all too common and management should ensure follow-up.

Reviewer 5: I should say ... management should ensure high priority of RCA action items and track follow through

Author: ```suggestion
Each post mortem feeds *extremely valuable* reliability improvements back to the team.
Your management team should ensure high priority of these actions & follow through on them.
```
@jeremyeder Added emphasis and wording about prioritising & following through.

#### Room to improve

Your team has sufficient time in their day to pursue improvement.
Your team needs room to practice processes, improve them and automate things.
Without this room, your team will mature more slowly and get bogged down in just keeping the service running day to day.
This can lead to scaling problems, both for the service and the team.
It isn't necessary to formally measure how much time is spent on keeping things running vs. feature development or automation.

> That can come later, however you are familiar with the concept of Toil.

Author: @mmazur sound reasonable for toil in phase 1?

## Phase 2: The ‘Walk’ Phase (WIP)

### Overview

Your team is well established, has working processes and is familiar with SRE principles.
A lot of those principles are baked into their processes & daily work.
In this phase your team is building on the foundations of observability and starting to reflect on their learnings about the service. 
Feedback loops, automation and 'shifting left' with different types of testing are the main theme of this phase.

> Your team is capable of scaling the service and number of users in a profitable way.

Reviewer 5: Your team is capable of scaling the service.  drop the last half

Author: ```suggestion
Your team is capable of scaling the service.
```

### Characteristics

* SLOs - SLOs have been iterated on, and you have more than 1 meaningful SLO. You understand and track error budgets
* Observability
    * Actionable Alerts
    * Self healing
        * Avoid SRE having sub-systems that automagically work around the inefficiencies of upstream - needs a good argument why it can’t be fixed upstream - pushing left
    * Combination of internal (whitebox) and external (blackbox) monitoring - tied to SLOs
    * Dashboards, logging (easily accessible)

> * Chaos Testing (possible candidate for Phase 1, but what would that look like? Ad-hoc testing?)

Author: @chaitanyaenr I'd appreciate your thoughts on where in the 3 phase maturity model you see chaos testing and chaos engineering

Reviewer 7: @david-martin think chaos testing/engineering should be part of all 3 phases but at different levels: 
Phase 1 - Understand how the service reacts to failures and feed the output into planning the environment accordingly while working on improving the areas that are weak

Phase 2 - Validate well established and iterated SLO's under turbulent conditions in addition to making sure right alerts are in place given that observability maturity is part of this phase

Phase 3 - Chaos testing in production during the available time window while still meeting the SLA.

Thoughts?

Author: This makes sense.
I've reworked the structure of the doc so that different areas have their own phases of maturity.
I've taken you're suggestions and worked them into a 'Failures' section

* Starting to get Exec support (improvment on 'buy-in')
* Start measuring Toil more formally & management track it
    * Develop relationship with the upstream (eng and sre, or sre & upstream product) for managing toil
* Improving the architecture for lower cost to serve (Is there a prereq here - calculating & knowing the cost to serve)
* DR & Incident Rehearsals (Identifying failure modes could be a prerequisite in phase 1)

> 

Reviewer 3: Could we add something here about "appropriate error logging". IMO, there is a habit used by devs to "chuck" all sorts of stuff into the logs (the old days of printf your output to debug!). They use it to help their development. I know because I have done this myself as a very foundational method :). This is fine for the first phase, but the remnants of these can hang around, even years later.
We have a mantra of: "Errors should actually be errors". The logs are the "UI" for SREs for the health of a service. If the logs are "polluted" by a whole lot of other non-errors, then that just makes it harder to resolve errors and increases the MTTR. I would also include in this the problem of not having errors where there should be an error.
Ideally, SRE is only notified when the actually is an error - but I'd say that is more for "phase 3".

Author: @brizrobbo Yes, definitely.
I'm thinking there's a lot there that could be captured as 'best practices' and linked off from the maturity model under the 'observability' headings in each phase.
As it stands I think https://issues.redhat.com/browse/SIGSRE-54 is where the logging best practices would be captured in such a doc (that issue may split as there's a lot in 'best practices').

I'm wondering how best to keep it sufficiently high level in these maturity phases and link off to additional info, guides, templates & worked examples where it makes sense.
In this case, I think saying something like this for phase 2?

```
You have established patterns and guidelines for observability signals such as metrics, logs, errors etc...
You are using these patterns to speed up debugging of issues and correlating between signals easier, thereby reducing your Mean Time To Recovery (MTTR).
Your alerts have matured and rarely give false positives as a result of having these patterns and refining queries.
```
How does that sound?

Phase 3 would then build on this foundation looking at AI Ops and improving the debug experience (focused on decreasing MTTR further)

Reviewer 2: The topic of logging also makes sense to mention in the operable software piece of the SIG's interests.

## Phase 3: The 'Run' Phase (WIP)

### Overview

> Your team has found a balance between keeping things running and building reliability into the service.

Reviewer 5: perhaps add something about having a roadmap, being proactive

Author: ```suggestion
Your team has found a balance between keeping things running and building reliability into the service.
This is evident by having a roadmap and a planning process for reliability features.
```

They have the ability to scale the service without having to increase team size.

> There is sufficient confidence in the service that your team looks at 'shifting right' with things like canary releases and production testing.

Reviewer 5: I am wondering if there will be a phase 4?  Because we should capture a phase that includes proficiency at these types of things.

Author: If there's enough meat in an area to have a 'doing it' and a 'proficient' phase, sure e.g. chaos testing vs. chaos engineering.
Alternatively they move to an earlier phase, but that may be overloading that phase. I've done that with SLOs & considering it for Chaos testing (pending feedback above).

I don't think I have sufficient knowledge captured on canary releases & production testing to make a decision here yet.
